src/useAsyncSetState.ts

```diff
@@ -0,0 +1,51 @@
+import React, { useCallback, useEffect, useRef, useState } from 'react'
+
+type Updater<TState> = (state: TState) => TState
+
+export type AsyncSetState<TState> = (
+  stateUpdate: React.SetStateAction<TState>,
+) => Promise<TState>
+
+export const useAsyncSetState = <TState>(
+  initialState: TState,
+): [TState, AsyncSetState<TState>] => {
+  const [state, setState] = useState(initialState)
+  const resolvers = useRef<((state: TState) => void)[]>([])
+
+  useEffect(() => {
+    resolvers.current.forEach(resolve => resolve(state))
+    resolvers.current = []
+  }, [state])
+
+  const setStateAsync = useCallback(
+    (update: Updater<TState> | TState) => {
+      return new Promise<TState>((resolve, reject) => {
+        setState(stateBefore => {
+          try {
+            let nextState: TState
+            if (typeof update === 'function') {
+              // @ts-ignore This isn't a Function
+              nextState = update(stateBefore)
+            } else {
+              nextState = update
+            }
+
+            // If state does not change, we must resolve the promise because
+            // react won't re-render and effect will not resolve
+            if (nextState === stateBefore) {
+              resolve(nextState)
```

Comment:
this doesn't seem right. suppose i run synchronously:

```js
const promise1 = setState(1);
const promise2 = setState(1);
const promise3 = setState(2);
```

`promise1` and `promise3` will resolve with `2`, while `promise2` will resolve with `1`.

Reply:
yeah not quite sure how to address that tho...I  guess check if there is a resolver queue already, only resolve early if there isn't any